### ISSUE TITLE
Find a non-google alternative to reCAPTCHA

### ISSUE THREAD

**Issue by richvdh** [matrix-org/matrix-doc#1281]
recaptcha is non-free and people would like to have non-google implementations.

**Comment by uhoreg** [comment]
see also https://github.com/vector-im/riot-web/issues/3606, which has some discussion of alternatives

**Comment by mt-dave** [comment]
I would think alternate of recaptcha will be kind of service, something that can solve traditional recaptcha issue like GDPR and accessibility and still provide solution like no captcha. 

I came across some solutions and here is a quick summary

 Captcha providers can widely be categorized in 2 categories :-

Captcha Service Providers : This option works well for mission critical Enterprises looking for protection against constantly evolving spam and bot threats.  Some of the Industry players in Captcha Services are :-

[RECAPTCHA](https://www.google.com/recaptcha/intro/v3.html)  : Free and One of the most widely used captcha service used across the globe. They have recently launched recaptcha v3  which generate a risk score based on user behavior on site, google cookies, traffic history etc.  GDPR has been a major concern considering what information it stores and uses for other google product like google ads.

[MTCaptcha ](https://mtcaptcha.com): Captcha Service that is more focused for Enterprise needs.  Provide NoCaptcha alternative to recaptcha, captcha account management, GDPR compliant, Availability across globe (China included).  Limited in low friction captcha capabilities.

Engagement Advertising Technology | CAPTCHA, Pre-roll, Brand Tags - Solve Media:  Ad driven Captcha that uses advertisement to generate captcha and solving them. GDPR compliant, Beautiful captcha and customizable. It may not be good idea to show advertisement on enterprise site.

Captcha Library Providers:  There are lot of players in Captcha Library space,  And if you are willing to manage and setup the code, some of the options are:-

BotDetect CAPTCHA Generator  : Most widely used captcha library,  Available in multiple languages. They license the library which then need to be implemented and managed.

KeyCAPTCHA - Innovative Anti-Spam Solution : Plugin driven captcha cover wide range of CMS systems. Mostly for CMS driven, need self hosting and management. Permutations are limited for captcha generation.

**Comment by richvdh** [comment]
I don't really know why it's being discussed there, since it's not specific to riot at all, but https://github.com/vector-im/riot-web/issues/3606 seems to be the authoritative issue on this.

**Comment by damnms** [comment]
imo its a combination of both
matrix-synapse by default supports google recaptcha, but nothing else, even tho it provides interfaces to implement any other captcha provider.
unfortunately, not every admin is programmer, so not any admin can implement his own captcha provider. therefore it would make sense to drop the support for google recaptcha in matrix and provide another default(!) captcha.

or: provide a simple documentation how to use matrix-synapse with any client AND another captcha provider.

**Comment by jryans** [comment]
Since this is a spec issue at it's core, I have closed https://github.com/vector-im/element-web/issues/3606 and redirected further discussion here.

**Comment by xaur** [comment]
Good idea to move the discussion here since the root of the problem is the [endorsement](https://github.com/matrix-org/matrix-doc/blob/d41ad7638e4f786c7fbbb51de17aade49b771c77/specification/client_server_api.rst#L729) of recaptcha at the spec level.

Would it be a too crazy request to remove Google recaptcha (and any proprietary anti-privacy fingerprinting service) from the spec?

**Comment by turt2live** [comment]
> Would it be a too crazy request to remove Google recaptcha (and any proprietary anti-privacy fingerprinting service) from the spec?

Not crazy, though without a proposed alternative it's unlikely to be accepted as an MSC. Having a captcha helps reduce a lot of spam accounts, particularly when paired with other registration requirements.

**Comment by Toby4213** [comment]
> > Would it be a too crazy request to remove Google recaptcha (and any proprietary anti-privacy fingerprinting service) from the spec?
> 
> Not crazy, though without a proposed alternative it's unlikely to be accepted as an MSC. Having a captcha helps reduce a lot of spam accounts, particularly when paired with other registration requirements.

Ok then I would strongly propose [hCaptcha](https://www.hcaptcha.com/) as an alternative to be added or completely replacing reCaptcha. Cloudflare changed their captcha provide to hCaptcha recently, see [here](https://blog.cloudflare.com/moving-from-recaptcha-to-hcaptcha/). I very much appreciate what Cloudflare does in general but especially regarding their [privacy policy](https://www.cloudflare.com/privacypolicy/). An endorsement by Cloudflare in my eyes is good enought for hCaptcha to be at leased supported by Matrix.

**Comment by turt2live** [comment]
The element-web issue has gone into quite a bit of discussion about the various captcha mechanisms. Realistically at this point someone needs to write a [proposal](https://matrix.org/docs/spec/proposals) for further discussion.

**Comment by Toby4213** [comment]
@turt2live ah ok thank you, well then I am going to write that proposal

**Comment by uhoreg** [comment]
> Besides, clearly there aren't very many options out there for one to implement anyway, which is why a decent one should be chosen and bundled with the clients.

From a *spec* perspective (i.e. the repo that this issue belongs to), we should allow any captcha.  Choosing a default one would be up to the homeservers (not the client), and once the spec allows any captcha, then you can argue about about which one to use in the repos for the homeservers.

**Comment by dev-love** [comment]
I think that FriendlyCaptcha might be the right solution. It's a proof-of-work based CAPTCHA alternative that respects the user's privacy. 

Their website address is https://friendlycaptcha.com and the front-end part is open source and available at https://github.com/friendlycaptcha/friendly-challenge. The European Union is using it on their official website (see https://www.eea.europa.eu/contact-us --> Ask your question)

What do you think about it?

**Comment by t3chguy** [comment]
20sec on a non flagship smartphone is enough time for someone to uninstall the app, same as grecaptcha basically.

**Comment by damnms** [comment]
why not leave that decision to the homeserver hoster.
if he'd like to use a slow but privacy-friendly alternative, he can decide whats best for him.
i would instantly switch to that, use it and see what goes on. 
registration is a one-time-thing. if a user is pissed because waiting for 20 seconds, he will be much more pissed because "no one is in matrix". at least imo and from my experience

**Comment by nicolas17** [comment]
Having *any* specific captcha service in the specification is absurd, so I don't know why people here are trying to pick an alternative. It should be way more flexible. The homeserver should give the client a URL to a webpage that uses whatever it wants to use to see if the user is legitimate or not.

The spec isn't even clear in *how* you're supposed to implement recaptcha. It only says you have to send a "captcha response" back. As a client implementor, how do I get that?

**Comment by ara4n** [comment]
Looks like we just got a PR at https://github.com/matrix-org/synapse/pull/8797 to implement hcaptcha.  (I'm a bit unclear on why hcaptcha is any better than recaptcha, given they're both proprietary centralised services, but yay for choice!)

**Comment by ara4n** [comment]
Sounds like

> Anyways, I'm pretty sure the right way to solve this issue is not to pick a new captcha system, but to fix the spec so that servers can use whatever captcha they want.

is going to be needed sooner than later :)

**Comment by Kegsay** [comment]
Dendrite will need to support a captcha system at some point and personally I'm leaning towards hcaptcha because:
 - The API is similar to reCAPTCHA.
 - It helps ML labelling which is nice.

It's more work to use a decentralised service but isn't something I'm opposed to.

**Comment by dev-love** [comment]
@Kegsay: I agree with your arguments and would be very happy if you considered our solution, Friendly Captcha, as well 👍 
Looking forward to your feedback! 😃 

> 
> 
> I think that FriendlyCaptcha might be the right solution. It's a proof-of-work based CAPTCHA alternative that respects the user's privacy.
> 
> Their website address is https://friendlycaptcha.com and the front-end part is open source and available at https://github.com/friendlycaptcha/friendly-challenge. The European Union is using it on their official website (see https://www.eea.europa.eu/contact-us --> Ask your question)

**Comment by Kegsay** [comment]
Friendly Captcha seems to have a 1k/monthly free tier limit which isn't ideal. Having it take 20s (!) on older mobiles is also not ideal, especially given we target mobiles of various ages. Twiddling thumbs for 20s feels a lot longer than solving a small labelling problem for 20s.

**Comment by nicolas17** [comment]
Sheesh, *don't tie the spec to any specific 3rd party service* :man_facepalming:

**Comment by turt2live** [comment]
The goal is to find a way to support arbitrary captcha, not to pick a specific provider (as mentioned already on this issue). The point of looking at providers anyways to figure out whether or not people would be likely to use them so they can be accommodated as best as possible in the spec, as it's just not feasible to support *every* provider. We can however try to support *most* or *popular* captcha providers.

**Comment by foresto** [comment]
Something worth keeping in mind while pondering designs:

It would be nice if signup process allowed for running a homeserver without a captcha. There are less intrusive ways to deal with automated signups, such as flagging them (based on behavior or inactivity) and deleting them later, and some admins might prefer to avoid things as user-hostile as captchas.

**Comment by turt2live** [comment]
@foresto that's already possible, and designers should be aware of that.

**Comment by grinapo** [comment]
Take your time, that's okay.
In the meantime go over to google search and type "recaptcha solver python", and enjoy the view.

The general problem seems to be that popular captchas soon have their solvers so the ideal fix for this specific problem is to provide some generic framework to use external captchas, be that one of the large ones (hcaptcha, recaptcha,... ), some more unusual ones (3d capctha, sketcha, ...) or housebrewn ones (like natural language questions/answers).

**Comment by bkil** [comment]
Do note that any CAPTCHA can be stepped through by simply piping through a CAPTCHA solver service a'la Amazon Mechanical Turk and thus paying a few cents for each registration. Not to you, to a third party and indirectly to your CAPTCHA provider in the form of completed free microwork. Let me not share link to these here.

A few alternatives come to mind:

* Require payment at registration, like a 1 cent bank transfer, 1 cent credit card lock & release or even via cryptocurrency.
* Google even supports _paying_ a few cents to each user during GCP registration to verify a unique bank account number.
* Sending or receiving an SMS
* As an alternative, back in the old free shared hosting days, there was a scheme where you only got an account that can be used for goodies after you proved to be a a nice citizen in some other platform.
* Web of trust. Making the platform invite-only, giving every user a certain amount of invites within a period of time. If you invite somebody who goes rogue, you would also get a negative reputation point (that decays with time). In case of abuse, the inviter could also be revealed within the affected community as a method of shaming. After a second or third strike, you get banned. Negative reputation might also be transitive with vanishing decaying factor as well. There should exist a better framework to detect sockpuppets based on IP & other fingerprints, though.

**Comment by turt2live** [comment]
(we don't really have an interest in making the protocol invite only or pay-to-access at this point, ftr)
